katcp/resource_client.py

```diff
@@ -1114,16 +1124,37 @@ def wait(self, sensor_name, condition_or_value, timeout=5):
             sensor is not present
 
         """
+        if quorum is None:
+            quorum = len(self.clients)
+        elif quorum < 1:
+            quorum = int(math.ceil(quorum * len(self.clients)))
+        if timeout and max_grace_period:
+            # Avoid having a grace period longer than or equal to timeout
+            grace_period = min(max_grace_period, timeout / 2.)
+            initial_timeout = timeout - grace_period
+        else:
+            grace_period = max_grace_period
+            initial_timeout = timeout
         # Build dict of futures instead of list as this will be easier to debug
         futures = {}
         for client in self.clients:
-            futures[client.name] = client.wait(sensor_name, condition_or_value,
-                                               timeout)
-        results = yield futures
+            f = client.wait(sensor_name, condition_or_value, initial_timeout)
+            futures[client.name] = f
+        # No timeout required here as all futures will resolve after timeout
+        initial_results = yield until_some(done_at_least=quorum, **futures)
+        results = dict(initial_results)
+        # Identify stragglers and let them all respond within grace period
+        stragglers = {}
+        for client in self.clients:
+            if not results.get(client.name, False):
+                f = client.wait(sensor_name, condition_or_value, grace_period)
+                stragglers[client.name] = f
+        rest_of_results = yield until_some(**stragglers)
+        results.update(dict(rest_of_results))
```

Comment:
Tornado does not support cancelling futures, so potentially you could be building up a bunch of futures that never resolve. In the case of using a coroutine (and some other mechanisms that I cannot remember off the top of my head), the ioloop itself will add a done callback to the future. This means that the ioloop (which is forever) has a reference to the future via the done callback. When a future is resolved it deletes all the done callbacks, so als deletes the ioloop -> future reference, allowing the future to be collected. If the future does _not_ resolve, it will not be collected.

We have in the past had an issue where we were waiting for a set of possible states in a loop using `until_any`() where each state was represented using a separate future. Each time the state waiting was done, a new set of futures were creates. When any of the states occured, the `until_any()` future would resolve, but not the states that were not reached. Those states were waited upon by the ioloop, so they would only be collected if they resolved. Some of the states happened very rarely, so copies of those futures built up with each iteration of the loop.

A bodgey hackey way of dealing with that was to have 'until_any()' reach into the future's private members and delete the  done  callbacks on the unresolved futures.

We _might_ run into a similar issue with `until_some()`, but since we are likely to use it only for things that are likely to happen within a fixed time and also in 'short-lived' observation scripts, it seemed better not to hack it. Hopefully tornado futures will one day allow themselves to be cancelled or we will switch over to another async framework :)